File: abc269/e.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;

int main()
{
    int N;
    cin >> N;
    //

    ll ok = 1;
    ll ng = N + 1;
    while (ng - ok > 0)
    {
        ll mid = (ok + ng) / 2;
        ll a = ok;
        ll b = mid;
        ll c = 1;
        ll d = N;
        cout << "? " << a << " " << b << " " << c << " " << d << endl;
        ll T;
        cin >> T;
        if (T == (b - a + 1))
        {
            ok = mid + 1;
        }
        else
        {
            ng = mid;
        }
    }
    auto u = ok;
    ok = 1;
    ng = N + 1;
    while (ng - ok > 0)
    {
        ll mid = (ok + ng) / 2;
        ll a = 1;
        ll b = N;
        ll c = ok;
        ll d = mid;
        cout << "? " << a << " " << b << " " << c << " " << d << endl;
        ll T;
        cin >> T;
        if (T == (d - c + 1))
        {
            ok = mid + 1;
        }
        else
        {
            ng = mid;
        }
    }
    auto l = ok;
    cout << "! " << u << " " << l << endl;
}

File: abc269/e.py
N = int(input())
l,r = 1,N+1
# ルークが置かれていない列を2部探索で求める

while r - l > 0:
    mid = (r+l) // 2
    A,B,C,D = 1,N,l,mid
    print(f'? {A} {B} {C} {D}')
    T = int(input())
    if T == (mid - l + 1): # ルークが置かれていない列は聞いた範囲ではない範囲に存在する
        l = mid+1
    else:
        r = mid

t,d = 1,N+1
while d - t > 0:
    mid = (t+d) // 2
    A,B,C,D = t,mid,1,N
    print(f'? {A} {B} {C} {D}')
    T = int(input())
    if T == (mid-t + 1): # ルークが置かれていない列は聞いた範囲ではない範囲に存在する
        t = mid + 1
    else:
        d = mid

print(f"! {t} {l}")


